### PR TITLE
Laboratory work #4, Yefimenko Natalya - 23FPL2

### PR DESCRIPTION
**Pull request by ShrubberyBush**

### CODE REVIEW

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -27,6 +29,10 @@ def get_paragraphs(text: str) -> list[str]:
         list[str]: Paragraphs from document.
     """
 
+    if not text:
+        raise ValueError
+    return text.split("\n")
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:34]: there can be blank strings after this operation, so result after parsing this input
```
hello, my friend!\n
\n
hello!
```
will be ['hello, my friend!','hello!']

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -41,6 +47,10 @@ def __init__(self) -> None:
         Initialize an instance of the BM25Vectorizer class.
         """
 
+        self._corpus = []
+        super().__init__(self._corpus)
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:51]: call super init at start of `__init__`

**Review by EIzharikov — COMMENTED** [review]
Ye, there are few comments, but dont relax! Make CI green and fix my comments. Good luck!

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (ShrubberyBush) [reply lab_4_retrieval_w_clustering/main.py:34]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (ShrubberyBush) [reply lab_4_retrieval_w_clustering/main.py:51]: Done

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -228,11 +341,17 @@ def set_new_centroid(self, new_centroid: Vector) -> None:
                 or if input arguments are empty.
         """
 
+        if not new_centroid:
+            raise ValueError
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: write error message here, please and in other errors

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -245,6 +364,11 @@ def add_document_index(self, index: int) -> None:
                 or if input arguments are empty.
         """
 
+        if index is None or not isinstance(index, int) or index < 0:
+            raise ValueError
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -272,11 +398,26 @@ def __init__(self, db: DocumentVectorDB, n_clusters: int) -> None:
...
+        self._n_clusters = n_clusters
+        self.__clusters = []
+
     def train(self) -> None:
         """
         Train k-means algorithm.
         """
 
+        if not self._db or not self._db.get_vectors():
+            raise ValueError
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -272,11 +398,26 @@ def __init__(self, db: DocumentVectorDB, n_clusters: int) -> None:
...
+        vectors = self._db.get_vectors()[:self._n_clusters]
+        if len(vectors) < self._n_clusters:
+            raise ValueError
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: .

**Review by EIzharikov — COMMENTED** [review]
Nothing special to review, just fix CI, please! In other case, your work seems stable, so good luck for the next review!

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (ShrubberyBush) [reply lab_4_retrieval_w_clustering/main.py]: All done

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -389,6 +644,10 @@ def make_report(self, num_examples: int, output_path: str) -> None:
             output_path (str): path to output file
         """
 
+        clusters_info = self.__algo.get_clusters_info(num_examples)
+        with open(output_path, "w", encoding="utf-8") as file:
+            json.dump(clusters_info, file)
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:651]: do you know the difference between ```dump``` and ```dumps```?

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -380,6 +616,25 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
             list[tuple[float, str]]: Relevant documents with their distances.
         """
 
+        if not query or not isinstance(query, str):
+            raise ValueError("Oops! Invalid input")
+        if not n_neighbours or not isinstance(n_neighbours, int) or n_neighbours <= 0:
+            raise ValueError("Oops! Invalid input")
+        query_token = self._db.get_tokenizer().tokenize(query)
+        if query_token is None:
+            raise ValueError("Oops! It's empty")
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: yes, it seems original, but for user it is useless input, do you understand?

Comment (EIzharikov) [reply lab_4_retrieval_w_clustering/main.py]: output message*

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -317,6 +502,30 @@ def get_clusters_info(self, num_examples: int) -> list[dict[str, int | list[str]
             list[dict[str, int| list[str]]]: List with information about each cluster
         """
 
+        if not isinstance(num_examples, int) or num_examples <= 0:
+            raise ValueError("Oops! Invalid input")
+        if not self.__clusters:
+            return []
+        clusters_info: list[dict[str, int | list[str]]] = []
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: no straight typings

**Review by EIzharikov — COMMENTED** [review]
So, your work is almost done, it is really great! Fix some my commentts and be ready for Friday, good luck!

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (ShrubberyBush) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (ShrubberyBush) [reply lab_4_retrieval_w_clustering/main.py:651]: `Dump` is used when the object needs to be stored in a file, while `dumps` is used when the object needs to be string

File: lab_4_retrieval_w_clustering/main.py (hunk 8 above)
Comment (ShrubberyBush) [reply lab_4_retrieval_w_clustering/main.py]: I guess so, fixed

File: lab_4_retrieval_w_clustering/main.py (hunk 9 above)
Comment (ShrubberyBush) [reply lab_4_retrieval_w_clustering/main.py]: Done

Comment (EIzharikov) [reply lab_4_retrieval_w_clustering/main.py]: 🏅

**Pull request closed without merging**